api/v1alpha1/nodepool_types.go

```diff
@@ -55,152 +55,281 @@ type NodePool struct {
+	// NodeCount is the latest observed number of nodes in the pool.
+	//
 	// +optional
 	NodeCount int32 `json:"nodeCount"`
 
-	// TODO (alberto): store list of existing nodes?
-	// Nodes []corev1.Nodes
-
-	Conditions []metav1.Condition `json:"conditions"`
-
-	// Version is the semantic version of the release applied by
-	// the hosted control plane operator.
-	// For a nodePool a given version represents the ignition config and
-	// an image artifact e.g an AMI in AWS.
+	// Version is the semantic version of the latest applied release specified by
+	// the NodePool.
+	//
 	// +kubebuilder:validation:Optional
 	Version string `json:"version,omitempty"`
+
+	// Conditions represents the latest available observations of the node pool's
+	// current state.
+	Conditions []metav1.Condition `json:"conditions"`
 }
 
-// +kubebuilder:object:root=true
 // NodePoolList contains a list of NodePools.
+//
+// +kubebuilder:object:root=true
 type NodePoolList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []NodePool `json:"items"`
 }
 
+// UpgradeType is a type of high-level upgrade behavior nodes in a NodePool.
 type UpgradeType string
 
-const UpgradeTypeInPlace = UpgradeType("InPlace")
-const UpgradeTypeReplace = UpgradeType("Replace")
+const (
+	// UpgradeTypeReplace is a strategy which replaces nodes using surge node
+	// capacity.
+	UpgradeTypeReplace = UpgradeType("Replace")
 
+	// UpgradeTypeInPlace is a strategy which replaces nodes in-place with no
+	// additional node capacity requirements.
+	UpgradeTypeInPlace = UpgradeType("InPlace")
+)
+
+// UpgradeStrategy is a specific strategy for upgrading nodes in a NodePool.
 type UpgradeStrategy string
 
-const UpgradeStrategyRollingUpdate = UpgradeStrategy("RollingUpdate")
-const UpgradeStrategyOnDelete = UpgradeStrategy("OnDelete")
+const (
+	// UpgradeStrategyRollingUpdate means use a rolling update for nodes.
+	UpgradeStrategyRollingUpdate = UpgradeStrategy("RollingUpdate")
 
+	// UpgradeStrategyOnDelete replaces old nodes when the deletion of the
+	// associated node instances are completed.
+	UpgradeStrategyOnDelete = UpgradeStrategy("OnDelete")
+)
+
+// ReplaceUpgrade specifies upgrade behavior that replaces existing nodes
+// according to a given strategy.
 type ReplaceUpgrade struct {
+	// Strategy is the node replacement strategy for nodes in the pool.
+	//
 	// +kubebuilder:validation:Optional
 	// +kubebuilder:validation:Enum=RollingUpdate;OnDelete
 	Strategy UpgradeStrategy `json:"strategy"`
+
+	// RollingUpdate specifies a rolling update strategy which upgrades nodes by
+	// creating new nodes and deleting the old ones.
+	//
 	// +kubebuilder:validation:Optional
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
 }
 
+// RollingUpdate specifies a rolling update strategy which upgrades nodes by
+// creating new nodes and deleting the old ones.
 type RollingUpdate struct {
```

Comment:
@enxebre I adapted these rolling update docs from the machine API, PTAL for accuracy

Reply:
lgmt, Thanks!